File: src/TypeChecker.cpp
#include "TypeChecker.h"

using namespace Types;

void TypeChecker::Visit(AST& node)
{
    // collect all top level functions, to not need forward declarations like in C
    m_findFunctions = true;
    for(auto* statement : node.statements)
        statement->Accept(this);


    m_findFunctions = false;
    for(auto* statement : node.statements)
        statement->Accept(this);
}

void TypeChecker::Visit(UnaryExpression& node)
{
    node.expr->Accept(this);
    if(!std::holds_alternative<Int>(m_currentType))
    {
        std::println(stderr, "{}: Unary expression must be of type int, got {}", node.expr->loc, m_currentType);
        m_error = true;
    }
    if(node.op == Op::NOT)
    {
        m_currentType = Bool();
    }
    else if(node.op == Op::U_MINUS)
    {
        m_currentType = Int();
    }
}

void TypeChecker::Visit(BinaryExpression& node)
{
    node.left->Accept(this);
    if(!std::holds_alternative<Int>(m_currentType))
    {
        std::println(stderr, "{}: Binary expression operands must be of type int, got {}", node.left->loc, m_currentType);
        m_error = true;
    }
    node.right->Accept(this);
    if(!std::holds_alternative<Int>(m_currentType))
    {
        std::println(stderr, "{}: Unary expression operands must be of type int, got {}", node.right->loc, m_currentType);
        m_error = true;
    }

    switch(node.op)
    {
    case Op::PLUS:
    case Op::MINUS:
    case Op::MUL:
    case Op::DIV:
    case Op::LSH:
    case Op::RSH:
    case Op::BAND:
    case Op::BOR:
        m_currentType = Int();
        break;

    case Op::LT:
    case Op::GT:
    case Op::LEQ:
    case Op::GEQ:
    case Op::EQUAL:
    case Op::NOT_EQUAL:
        m_currentType = Bool();
        break;
    }
}

void TypeChecker::Visit(NumberLiteral& node)
{
    m_currentType = Int();
}

void TypeChecker::Visit(StringLiteral& node)
{
    m_currentType = String();
}
void TypeChecker::Visit(BoolLiteral& node)
{
    m_currentType = Bool();
}

void TypeChecker::Visit(VariableAccess& node)
{
    if(node.index)
    {
        if(std::holds_alternative<Array>(m_stack.Find(node.name, node.loc)))
        {
            m_currentType = std::get<Array>(m_stack.Find(node.name, node.loc)).type[0];
        }
        else if(std::holds_alternative<String>(m_stack.Find(node.name, node.loc)))
        {
            m_currentType = Int();
        }
        else
        {
            std::println(stderr, "{}: Expected array or string type for variable access but got {}", node.loc, m_stack.Find(node.name, node.loc));
            m_error = true;
        }
    }
    else
    {
        m_currentType = m_stack.Find(node.name, node.loc);
    }
}

void TypeChecker::Visit(VariableAssignment& node)
{
    node.value->Accept(this);

    Type t;
    if(node.index)
    {
        t = std::get<Array>(m_stack.Find(node.name, node.loc)).type[0];
    }
    else
    {
        t = m_stack.Find(node.name, node.loc);
    }
    if(m_currentType.index() != t.index())
    {
        std::println(stderr, "{}: Expected {} type for variable assignment but got {}", node.value->loc, t, m_currentType);
        m_error = true;
    }
}

void TypeChecker::Visit(Block& node)
{
    for(auto* statement : node.statements)
        statement->Accept(this);
    if(node.expr)
        node.expr->Accept(this);
}

void TypeChecker::Visit(IfElse& node)
{
    node.condition->Accept(this);
    if(std::holds_alternative<Bool>(m_currentType) == false)
    {
        std::println(stderr, "{}: Expected bool type for if condition but got {}", node.condition->loc, m_currentType);
        m_error = true;
    }
    node.body->Accept(this);
    if(node.elseBlock)
    {
        Type bodyType = m_currentType;
        node.elseBlock->Accept(this);

        if(bodyType.index() != m_currentType.index())
        {
            std::println(stderr, "{}: The body of both branches of an if must have the same type. Type of if branch: {}, type of else branch: {}", node.body->loc, bodyType, m_currentType);
            m_error = true;
        }
    }
}

void TypeChecker::Visit(WhileLoop& node)
{
    node.condition->Accept(this);
    if(std::holds_alternative<Bool>(m_currentType) == false)
    {
        std::println(stderr, "{}: Expected bool type for while condition but got {}", node.condition->loc, m_currentType);
        m_error = true;
    }
    node.body->Accept(this);
    if(std::holds_alternative<NoneType>(m_currentType) == false)
    {
        std::println(stderr, "{}: While loop's body must have NoneType, but got {}", node.body->loc, m_currentType);
        m_error = true;
    }
}

void TypeChecker::Visit(FnCall& node)
{
    Function funType;
    try
    {
        funType = std::get<Function>(m_stack.Find(node.name, node.loc));
    }
    catch(std::bad_variant_access&)
    {
        std::println(stderr, "{}: Expected function type for function call but got {}", node.loc, m_currentType);
        m_error = true;
    }

    if(funType.parameters.size() != node.arguments.size())
    {
        std::println(stderr, "{}: Expected {} arguments for function call but got {}", node.loc, funType.parameters.size(), node.arguments.size());
        m_error = true;
    }

    for(int i = 0; i < node.arguments.size() && i < funType.parameters.size(); i++)
    {
        auto* arg = node.arguments[i];
        arg->Accept(this);
        if(m_currentType.index() != funType.parameters[i].index())
        {
            std::println(stderr, "{}: Expected {} type for function argument but got {}", arg->loc, funType.parameters[i], m_currentType);
            m_error = true;
        }
    }
    m_currentType = m_stack.Find(node.name, node.loc);
}

void TypeChecker::Visit(ExpressionStatement& node)
{
    if(m_findFunctions)
    {
        return;
    }
    node.expr->Accept(this);
    m_currentType = NoneType();  // expression statements just ignore the type of the expression
}

void TypeChecker::Visit(VariableDeclaration& node)
{
    if(m_findFunctions)
    {
        return;
    }

    if(node.value)
    {
        node.value->Accept(this);

        if(m_currentType.index() != node.type.index())
        {
            std::println(stderr, "{}: Expected {} type for variable initialisation but got {}", node.value->loc, node.type, m_currentType);
            m_error = true;
        }
    }
    m_stack.Push(node.name, node.type);

    m_currentType = NoneType();
}

void TypeChecker::Visit(FnDeclaration& node)
{
    m_stack.Push(node.name, node.type);
    if(m_findFunctions)
    {
        return;
    }

    for(int i = 0; i < node.type.parameters.size(); i++)
    {
        m_stack.Push(node.parameters[i], node.type.parameters[i]);
    }
    node.body->Accept(this);

    if(m_currentType.index() != node.type.returnType[0].index())
    {
        Location loc = node.loc;
        if(node.body->expr)
        {
            loc = node.body->expr->loc;
        }
        else if(!node.body->statements.empty())
        {
            loc = node.body->statements.back()->loc;
        }
        std::println(stderr, "{}: Expected {} for function return value type but got {}", loc, node.type.returnType[0], m_currentType);
        m_error = true;
    }


    m_currentType = NoneType();
}

File: lib/type_checker.ml
open Core
open Ast
module H = Hashtbl.Make (String)

exception SymbolNotFound of loc * name
exception FunctionRedefinition of loc * name
exception InvalidType of loc * var_type * var_type
exception InvalidReturnType of loc * var_type * var_type
exception BranchTypeMismatch of loc * var_type * var_type
exception InvalidArgumentNum of loc * int * int

type env = {
    parent : env option;
    symbols : var_type H.t;
  }

let rec find_type loc sym env =
    match Hashtbl.find env.symbols sym with
    | Some t -> t
    | None -> (
        match env.parent with
        | Some parent -> find_type loc sym parent
        | None -> raise (SymbolNotFound (loc, sym)))

let add_symbol env name typ = Hashtbl.set env.symbols ~key:name ~data:typ

let check ast =
    let rec check_statement node env =
        match node.node with
        | Declaration (name, t) ->
            add_symbol env name t;
            env
        | Declaration_assign (name, t, expr) ->
            let assign_type = check_expr expr env in
            if equal_var_type assign_type t then (
              add_symbol env name t;
              env)
            else
              raise (InvalidType (expr.loc, t, assign_type))
        | FnDeclaration (name, t, param_names, body) -> (
            (* fn definition only in global scope for now *)
            assert (Option.is_none env.parent);
            match t with
            | Fn (ret, param_types) ->
                if Hashtbl.mem env.symbols name then
                  raise (FunctionRedefinition (node.loc, name))
                else
                  let body_env = { parent = Some env; symbols = H.create () } in
                  List.iter2_exn param_names param_types ~f:(fun id typ ->
                      add_symbol body_env id typ);
                  let body_type = check_expr body body_env in
                  if equal_var_type ret body_type then (
                    add_symbol env name t;
                    env)
                  else
                    raise (InvalidReturnType (node.loc, ret, body_type))
            | _ -> assert false)
        | ExprStatement expr ->
            check_expr expr env |> ignore;
            env
        | While (cond, body) ->
            let cond_type = check_expr cond env in
            if Ast.equal_var_type cond_type (Type "bool") then
              let body_type = check_expr body env in
              if Ast.equal_var_type body_type (Type "void") then
                env
              else
                raise (InvalidType (node.loc, Type "void", body_type))
            else
              raise (InvalidType (node.loc, Type "bool", cond_type))
    and check_expr node env =
        match node.node with
        | String _ -> Type "str"
        | Int _ -> Type "int"
        | Bool _ -> Type "bool"
        | Nullptr -> Type "nullptr"
        | Variable (name, index_expr) -> (
            let typ =
                match find_type node.loc name env with
                | Ast.Array (t, _) -> t
                | t -> t
            in
            match index_expr with
            | None -> typ
            | Some e -> (
                match check_expr e env with
                | Type "int" -> typ
                | t -> raise (InvalidType (e.loc, Type "int", t))))
        | Add (lhs, rhs)
        | Sub (lhs, rhs)
        | Mul (lhs, rhs)
        | Div (lhs, rhs)
        | Lsh (lhs, rhs)
        | Rsh (lhs, rhs)
        | BAnd (lhs, rhs)
        | BOr (lhs, rhs) -> (
            match check_expr lhs env with
            | Type "int" -> (
                match check_expr rhs env with
                | Type "int" -> Type "int"
                | t -> raise (InvalidType (rhs.loc, Type "int", t)))
            | t -> raise (InvalidType (lhs.loc, Type "int", t)))
        | Gt (lhs, rhs)
        | GEq (lhs, rhs)
        | Lt (lhs, rhs)
        | LEq (lhs, rhs)
        | Eq (lhs, rhs)
        | NEq (lhs, rhs) -> (
            match check_expr lhs env with
            | Type "int" -> (
                match check_expr rhs env with
                | Type "int" -> Type "bool"
                | t -> raise (InvalidType (rhs.loc, Type "int", t)))
            | t -> raise (InvalidType (lhs.loc, Type "int", t)))
        | UNot e -> (
            match check_expr e env with
            | Type "bool" -> Type "bool"
            | t -> raise (InvalidType (e.loc, Type "bool", t)))
        | VarAssign (name, expr) ->
            let t = find_type node.loc name env in
            let assign_type = check_expr expr env in
            if equal_var_type assign_type t then
              t
            else
              raise (InvalidType (expr.loc, t, assign_type))
        | ArrayVarAssign (name, index_expr, expr) -> (
            let t =
                match find_type node.loc name env with
                | Ast.Array (t, _) -> t
                | _ -> assert false
            in
            match check_expr index_expr env with
            | Type "int" ->
                let assign_type = check_expr expr env in
                if equal_var_type assign_type t then
                  t
                else
                  raise (InvalidType (expr.loc, t, assign_type))
            | x -> raise (InvalidType (index_expr.loc, Type "int", x)))
        | IfElse (cond, body, else_body) -> (
            match check_expr cond env with
            | Type "bool" ->
                let body_typ = check_expr body env in
                let else_typ =
                    match else_body with
                    | Some e -> check_expr e env
                    | None -> Type "void"
                in
                if equal_var_type body_typ else_typ then
                  body_typ
                else
                  raise (BranchTypeMismatch (node.loc, body_typ, else_typ))
            | t -> raise (InvalidType (cond.loc, Type "bool", t)))
        | Block (stmt_list, e) -> (
            let new_env = { parent = Some env; symbols = H.create () } in
            let new_env =
                List.fold stmt_list ~init:new_env ~f:(fun env s -> check_statement s env)
            in
            match e with
            | None -> Type "void"
            | Some e -> check_expr e new_env)
        | FnCall (name, args) -> (
            match find_type node.loc name env with
            | Fn (ret, params) ->
                let rec aux args_list params_list num_args num_params =
                    match (args_list, params_list) with
                    | [], [] -> ret
                    | arg :: at, param :: pt ->
                        let arg_type = check_expr arg env in
                        if equal_var_type arg_type param then
                          aux at pt (num_args + 1) (num_params + 1)
                        else
                          raise (InvalidType (arg.loc, param, arg_type))
                    | l, [] ->
                        raise (InvalidArgumentNum (node.loc, num_params, num_args + List.length l))
                    | [], l ->
                        raise (InvalidArgumentNum (node.loc, num_params + List.length l, num_args))
                in
                aux args params 0 0
            | t -> raise (InvalidType (node.loc, Fn (Type "fn", []), t)))
    in
    let env0 = { parent = None; symbols = H.create () } in
    add_symbol env0 "print" (Fn (Type "void", [ Type "str" ]));
    add_symbol env0 "print_int" (Fn (Type "void", [ Type "int" ]));
    try
      List.fold ast ~init:env0 ~f:(fun env s -> check_statement s env) |> ignore;
      true
    with
    | InvalidType (loc, expected, got) ->
        Printf.eprintf "%s:%d:%d: Type error: expected `%s` but got `%s`.\n" loc.filename loc.line
          loc.col (show_var_type expected) (show_var_type got);
        false
    | InvalidReturnType (loc, expected, got) ->
        Printf.eprintf "%s:%d:%d: Type error: expected function to return `%s` but got `%s`.\n"
          loc.filename loc.line loc.col (show_var_type expected) (show_var_type got);
        false
    | BranchTypeMismatch (loc, body_typ, else_typ) ->
        Printf.eprintf
          "%s:%d:%d: Type error: The two branches of an if expression must have the same type got \
           `%s` and `%s`.\n"
          loc.filename loc.line loc.col (show_var_type body_typ) (show_var_type else_typ);
        false
    | InvalidArgumentNum (loc, expected, got) ->
        Printf.eprintf "%s:%d:%d: Error: expected %d arguments but got %d.\n" loc.filename loc.line
          loc.col expected got;
        false
    | SymbolNotFound (loc, name) ->
        Printf.eprintf "%s:%d:%d: Undefined symbol `%s`.\n" loc.filename loc.line loc.col name;
        false
    | FunctionRedefinition (loc, name) ->
        Printf.eprintf "%s:%d:%d: Redefinition of function `%s`.\n" loc.filename loc.line loc.col
          name;
        false
